### PR TITLE
Add Metadata and UserDefinedMetadata to DataElement.

### PR DESCRIPTION
**Pull request by bjorntore**
DIBK feature request:

- Ability to set a list of key-value pairs on data elements, much like tags, but they need a key to go with the value to identify what the value is later.
- The ability to use the AllowedContributors property on DataType to authorize editing of the key-value list. We only used this concept one other place (create data element), but we decided to use it one more place, at least until we support proper authorization on the data type level.
- Duplication check on keys.
- Wanted the ability to specify which keys are allowed to be used on a certain DataType.
- This key-value list is named UserDefinedMetadata and has an endpoint for editing it, so the end user can change it freely. We also decided to add another key-value list for "internal use" in app code, which the end user can not edit called Metadata.

Most of the UserDefinedMetadata functionality is in the controller. It started very simple, but got a more complex towards the end. Should I stick it in a service?

## Related Issue(s)
- #392

## Verification
- [x] **Your** code builds clean without any errors or warnings
- [x] Manual testing done (required)
- [x] Relevant automated test added (if you find this hard, leave it and we'll help out)
- [x] All tests run green

## Documentation
- [ ] User documentation is updated with a separate linked PR in [altinn-studio-docs.](https://github.com/Altinn/altinn-studio-docs) (if applicable)

### CODE REVIEW

File: src/Altinn.App.Api/Controllers/UserDefinedMetadataController.cs [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+    /// <param name="org">The short name for the application owner.</param>
+    /// <param name="app">The name of the application.</param>
+    /// <param name="instanceOwnerPartyId">The party id of the owner of the instance.</param>
+    /// <param name="instanceGuid">The id of the instance.</param>
+    /// <param name="dataGuid">The id of the data element.</param>
+    /// <returns>A <see cref="UserDefinedMetadataDto"/> object with a list of key value entries.</returns>
+    [HttpGet]
+    [Authorize(Policy = AuthzConstants.POLICY_INSTANCE_READ)]
+    [ProducesResponseType(typeof(UserDefinedMetadataDto), StatusCodes.Status200OK)]
+    public async Task<ActionResult<UserDefinedMetadataDto>> Get(
```
[the remark this answers is not in the document]
Comment (bjorntore) [reply src/Altinn.App.Api/Controllers/UserDefinedMetadataController.cs:62]: Fixed

File: src/Altinn.App.Api/Controllers/UserDefinedMetadataController.cs [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+    /// <param name="app">The name of the application.</param>
+    /// <param name="instanceOwnerPartyId">The party id of the owner of the instance.</param>
+    /// <param name="instanceGuid">The id of the instance.</param>
+    /// <param name="dataGuid">The id of the data element.</param>
+    /// <param name="userDefinedMetadataDto">Object with a list of properties.</param>
+    /// <returns>A <see cref="UserDefinedMetadataDto"/> object with a list of key value entries.</returns>
+    [HttpPut]
+    [Authorize(Policy = AuthzConstants.POLICY_INSTANCE_WRITE)]
+    [ProducesResponseType(typeof(UserDefinedMetadataDto), StatusCodes.Status200OK)]
+    public async Task<ActionResult<UserDefinedMetadataDto>> Update(
```
[the remark this answers is not in the document]
Comment (bjorntore) [reply src/Altinn.App.Api/Controllers/UserDefinedMetadataController.cs:98]: Fixed

File: src/Altinn.App.Api/Helpers/ValidContributorHelper.cs [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+/// Helper class for validating if a user is a valid contributor to a data type.
+/// </summary>
+/// <remarks>
+/// TODO: The concept of inline authorization of valid contributors is not widely used and is likely not the best approach for doing authorization on the data type level, but there is no support for it yet in the policy based authorization, so keeping for now.
+/// </remarks>
+internal static class ValidContributorHelper
+{
+    internal static bool IsValidContributor(DataType dataType, string? org, int? orgNr)
+    {
+        if (dataType.AllowedContributers == null || dataType.AllowedContributers.Count == 0)
```
Comment (HauklandJ) [inline src/Altinn.App.Api/Helpers/ValidContributorHelper.cs]: nit: prefer 'is null' to '== null'

**Review by HauklandJ — COMMENTED** [review]
Looks good to me structurally and logically. I do however not see where the extension method AllowedKeysForUserDefinedMetadata on DataType is defined. The build fails for this reason locally

**Review by HauklandJ — APPROVED** [review]
lgtm

**Pull request merged**